### DISCUSSION TITLE
TestContext.Progress and test log/output

### DISCUSSION THREAD

**Discussion by dfev77** [nunit/nunit#4287] [Q&A] [answered]
Hi,

I recently started to use `TestContext.Progress` as a way to track progress of some long running tests so that I can stop them early if something does not work as expected. And it works fine.
I've been replacing a couple of `TestContent.Writeline` with `TestContext.Progress.WriteLine` only to realize that the Progress messages are not written to the test output or TestResults.xml, making tracing the execution **after the run** impossible.

Is there a way to have `TestContext.Progress` write to both "progress" and "out" streams? (other then writing to both streams my self)

Thanks,
florin

**Comment by stevenaw** [comment, marked as answer]
Hi @dfev77 
To my knowledge they are different output streams to serve different use cases, but if you want to write to both easily you would likely need to code that yourself. A minimal way to do this could be define your own extension method on the `TestContext` class. For example, the following untested code:

```csharp
public static class TestContextExtensions {
  public static void LogToProgressAndOutput(this TestContext context, string message) {
    context.Out.WriteLine(message);
    context.Progress.WriteLine(message);
  }
}
```

could then be invoked as

```csharp
TestContext.Current.LogToProgressAndOutput("hello world");
```

**Reply by dfev77** [reply]
I would debate on different purposes of the two streams, but I suppose my usage might be a corner case, which does not fit the NUnit way of monitoring execution and analyzing test results. 
When you want to later analyze the execution of a test you are likely to want the "progress" messages, which are likely to be something like "start long running operation 1", in the overall output of the test.

Thanks for your input.
